### ISSUE TITLE
Generator may fail with "Could not load file or assembly 'YamlDotNet..."

### ISSUE THREAD

**Issue by AraHaan** [microsoft/CsWin32#221]
Currently for some reason the source generator starts to bail out when it sees ``GetCurrentProcess`` when called inside of ``dotnet build``.

Oddly it seems to generate just fine from within Visual Studio even if I use the .NET 6 Preview 1/2 SDK's.

```
CSC : warning CS8785: Generator 'SourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'GenerationFailedException' with message 'Failed while generating extern method: GetCurrentProcess' [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MiniDump.cs(15,21): error CS0234: The type or namespace name 'Windows' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?) [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(9,21): error CS0234: The type or namespace name 'Windows' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?) [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MiniDump.cs(88,132): error CS0246: The type or namespace name 'MINIDUMP_EXCEPTION_INFORMATION' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MiniDump.cs(88,180): error CS0246: The type or namespace name 'MINIDUMP_USER_STREAM_INFORMATION' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MiniDump.cs(88,231): error CS0246: The type or namespace name 'MINIDUMP_CALLBACK_INFORMATION' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(19,18): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(25,24): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(31,26): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(37,26): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(43,24): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(49,22): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(56,31): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(63,42): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(70,29): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(77,33): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(84,38): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(91,31): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(98,30): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(105,26): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(112,24): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(115,33): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(121,34): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(128,38): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(135,36): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(142,32): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(149,29): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(156,24): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(159,32): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(162,24): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(165,40): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump\MinidumpTypes.cs(170,26): error CS0103: The name 'MINIDUMP_TYPE' does not exist in the current context [C:\Users\User\Desktop\github\MyPackages\MiniDump\MiniDump.csproj]
```

This happens with my project with the following ``NativeMethods.txt``:

```
GetCurrentProcess
GetCurrentProcessId
GetCurrentThreadId
MiniDumpWriteDump
```

The first 2 is used when calling ``MiniDumpWriteDump`` and the 3rd one is for when constructing the ``EXCEPTION_INFORMATION`` structure for the call.

and the following ``NativeMethods.json`` file:
```json
{
  "$schema": "https://raw.githubusercontent.com/microsoft/CsWin32/main/src/Microsoft.Windows.CsWin32/settings.schema.json",
  "allowMarshaling": false
}
```

**Comment by AraHaan** [comment]
I *could* use ``System.Diagnostics.Process`` however the issue is that it does not:
- return the process ids as uint's.
- Seems a bit useless if I get the current process from ``System.Diagnostics.Process`` only to use the handle on the call.

**Comment by jnm2** [comment]
@AraHaan If you are using .NET 5+, you should be using `https://docs.microsoft.com/en-us/dotnet/api/system.environment.processid`.

**Comment by AraHaan** [comment]
I use the .NET 6 SDK but on that project I target ``netstandard2.0``.

**Comment by AraHaan** [comment]
It also seems to do it with ``MessageBoxW`` as well when on another project.

**Comment by AraHaan** [comment]
Hmm it does it on every Windows API function when targeting .NET Standard 2.0.

**Comment by AraHaan** [comment]
@AArnott Any idea why it wont at least let the generations of the enums when using the SourceGenerator on ``netstandard2.0`` targets? Enums and structs at *least* (if not the actual p/invokes) should be safe for .NET Standard 2.0.

**Comment by AraHaan** [comment]
Also I feel like the warning it prints should have included the stack trace of the exception that caused the ``GenerationFailedException`` exception so that way I know where something would need a patch on the source generator.

**Comment by AArnott** [comment]
It doesn't repro for me on 0.1.422-beta (latest on nuget.org) either.

**Comment by jnm2** [comment]
Could this be a result of combining a particular version of win32metadata?

**Comment by AArnott** [comment]
@jnm2 potentially yes. We only support using CsWin32 with the metadata that it brings in automatically, because the metadata churns so much right now. Updating metadata on your own is taking on risk that things break.
@AraHaan please let us know if you're using a non-default metadata version as well.

**Comment by AraHaan** [comment]
I was using the 0.1.422-beta from nuget.org, but even without the metadata on my end every version kept failing for it (until I copied the contents of the file and added a define to turn off the copy for now to sync it.

However it is a hack and it's also been pushed to my repository as well too: https://github.com/Elskom/Sdk/tree/use-sourcegenerators/MiniDump
I had to paste the file in to: https://github.com/Elskom/Sdk/blob/use-sourcegenerators/MiniDump/MiniDump/PInvokes/MINIDUMP_TYPE.cs but it could be deleted and checked on another pc to see if it works (for reproducibility).

Perhaps something in the project file is causing it.

**Comment by AraHaan** [comment]
I have just force pushed to that branch and so when you try to do ``dotnet build --configuration Release`` on Windows, Mac, Linux, whatever OS, it will fail on the source generator either way with this issue (and yes it removes the metadata package entirely and defines that ``ENUM_SYNC`` so then it would try to generate that enum, but fail).

**Comment by AraHaan** [comment]
Ok, I have simplified the reproducibility of this issue down to clone ``git clone https://github.com/Elskom/Sdk.git --recursive --branch use-sourcegenerators`` followed by a build``dotnet build --configuration Release`` 😂.

**Comment by AraHaan** [comment]
I decided I wanted to test it on github actions using ``ubuntu-latest`` and I found it also reproduces there too 😂 https://github.com/Elskom/Sdk/pull/190/checks?check_run_id=2410109760.

**Comment by AraHaan** [comment]
I tested locally a change and still could not fix it.

**Comment by AArnott** [comment]
I can repro the exception to generate `MiniDumpWriteDump` at the tip of main. I'll investigate.

**Comment by AArnott** [comment]
Well, the repro I'm seeing is unlikely to be related to what you were seeing, since it's a new failure from #270. But I have a PR to improve the experience either way so we can better diagnose such failures in the future.

**Comment by AraHaan** [comment]
Also mind checking on that fix to see if the other pr I made that adds tests for those MiniDump api calls and see if they get generated for ``netstandard2.0``?

**Comment by AraHaan** [comment]
```warning CS8785: Generator 'SourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'GenerationFailedException' with message 'Failed to generate MINIDUMP_TYPE'```

I tried copiling again with ``0.1.474-beta`` and still get this for that enum. I do not see any places where the data in that enum is not consistent across all platforms of Windows. I think that enum should be emitted for AnyCPU.

**Comment by AArnott** [comment]
The `MINIDUMP_TYPE` enum is indeed available for AnyCPU generation. I still have no idea why it's failing. Tonight's CI will have a better error message to explain it, but in the meantime if you share a repro I can take a look.

**Comment by AraHaan** [comment]
Currently it is reproducible with ``git clone https://github.com/Elskom/Sdk.git --recursive --branch use-sourcegenerators`` and then ``dotnet build --configuration Release``.

**Comment by AArnott** [comment]
Great. With those repro steps I can see this error with the latest CsWin32 version:
> error PInvoke000: An internal error occurred: Failed to generate MINIDUMP_TYPE. The type initializer for 'Microsoft.Windows.CsWin32.Docs' threw an exception. Could not load file or assembly 'YamlDotNet, Version=11.0.0.0, Culture=neutral, PublicKeyToken=ec19458f3c15af5e'. Could not find or load a specific file. (0x80131621). [C:\temp\Sdk\MiniDump\MiniDump.csproj]

So this may be a duplicate of #261, which is probably not completely fixed. YamlDotNet 11.0.0.0 is certainly included in the package, so it is probably some dependency whose assembly version is not precisely satisfied.

**Comment by AArnott** [comment]
Ok, by using a debugger I found that YamlDotNet.dll has already been loaded from %userprofile%\.nuget\packages\securitycodescan.vs2019\5.1.0\analyzers\dotnet\YamlDotNet.dll

I don't know what securitycodescan.vs2019 is, but this package includes YamlDotNet v8.1.2.0 which loads first and therefore causes .NET Core to be unwilling to load the 11.0.0.0 used by CsWin32 later in the build. 
I therefore do not expect this to be a problem to hit most people.

I am concerned however at the lack of isolation between analyzers that the C# compiler IMO ought to be providing so that one analyzer's dependencies does not cause problems for another's dependencies. I'll follow up with the Roslyn team on this.

**Comment by AArnott** [comment]
@arahaan, when I remove 

```xml
    <PackageReference Include="SecurityCodeScan.VS2019" IsImplicitlyDefined="true" Version="*-*">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
```
from your Sdk.props file, the CsWin32 error goes away (and is replaced by others that appear unrelated).

I don't know if dropping that dependency for now is an option for you though. I've already reached out to the Roslyn team to ask for recommendations.

**Comment by AraHaan** [comment]
Thanks so the issue is a combination of things. Wonder what other Analyzers that I use use a different version of ``YamlDotNet`` causing this conflict.

**Comment by JarLob** [comment]
> I am concerned however at the lack of isolation between analyzers that the C# compiler IMO ought to be providing so that one analyzer's dependencies does not cause problems for another's dependencies. I'll follow up with the Roslyn team on this.

Hey @AArnott, I'm interested in hearing what is the take of the Roslyn team on it. Is there a link? Is it public issue?

**Comment by AArnott** [comment]
Here is what I heard back from the Roslyn team:

> Essentially the build works when you use msbuild but not “dotnet build”. The latter is happening on .NET Core and we don’t fully isolate anaylzers in independent ALC yet. It’s on the backlog for Dev17 / C# 10.

**Comment by jnm2** [comment]
Is that this issue? https://github.com/dotnet/roslyn/issues/52177

**Comment by AraHaan** [comment]
It looks like it.

**Comment by JarLob** [comment]
Does `/p:BuildInParallel=false` as described workarounds the issue in your case?

**Comment by AraHaan** [comment]
> Does `/p:BuildInParallel=false` as described avoids the issue in your case?

I tried that and this was the result:
```
NativeMethods.txt(1,1): error PInvoke000: An internal error occurred: Failed to generate MINIDUMP_TYPE. The type initializer for 'Microsoft.Windows.CsWin32.Docs' threw an exception. Could not load file or assembly 'YamlDotNet, Version=11.0.0.0, Culture=neutral, PublicKeyToken=ec19458f3c15af5e'. Could not find or load a specific file. (0x80131621). Could not load file or assembly 'YamlDotNet, Version=11.0.0.0, Culture=neutral, PublicKeyToken=ec19458f3c15af5e'.
```

**Comment by AraHaan** [comment]
Hmm I wonder if all analyzers should start doing binding redirects of all of their dependencies from ``0.0.0.0-65535.65535.65535.65535`` (aka up to whatever the max valid possible version number is) for the ``oldVersion`` element, and as for ``newVersion`` not sure if that can be wildcarded to ``*`` and then have whatever the first version of said dependency to be loaded to be the one that wins the war.

**Comment by jnm2** [comment]
@JarLob `/p:BuildInParallel=false` only worked for me as an accident. It's the explanation for why `msbuild` always worked and `dotnet build` often failed. But the fact that things were loading in the right order was specific to my repro; I think in this case, they load in the wrong order.

**Comment by AArnott** [comment]
It isn't parallelism that breaks it. It may simply be a build ordering issue. And it's super complex for the roslyn team to fix. It's not likely something we can totally workaround within the source generator, although we may be able to improve some scenarios.

**Comment by AArnott** [comment]
The roslyn issue tracking it is https://github.com/dotnet/roslyn/issues/53672

**Comment by AArnott** [comment]
I'm blocked by this myself in https://github.com/dotnet/Nerdbank.GitVersioning/pull/616. `msbuild.exe` and VS work but `dotnet build` fails.

**Comment by AraHaan** [comment]
@AArnott what if the part of the generator that uses YamlDotNet was split into a console program that the generator ships with and somehow is able to execute in a subprocess of the compiler to work around this?

Is it possible to do that and permanently fix the YamlDotNet issue for now?

**Comment by AArnott** [comment]
We use `YamlDotNet` to deserialize a file into a memory model we need. Moving it out of proc would defeat the point.
However as part of fixing #38 we *may* remove use of YamlDotNet anyway.

**Comment by AArnott** [comment]
We did indeed drop the YamlDotNet dependency.
